timemachine/cpp/src/nonbonded_all_pairs.cu

```diff
@@ -123,21 +131,60 @@ template <typename RealType> NonbondedAllPairs<RealType>::~NonbondedAllPairs() {
 
 template <typename RealType> void NonbondedAllPairs<RealType>::set_nblist_padding(double val) { nblist_padding_ = val; }
 
+template <typename RealType> void NonbondedAllPairs<RealType>::verify_atom_idxs(const std::vector<int> &atom_idxs) {
+    if (atom_idxs.size() == 0) {
+        throw std::runtime_error("idxs can't be empty");
+    }
+    std::set<int> unique_idxs(atom_idxs.begin(), atom_idxs.end());
+    if (unique_idxs.size() != atom_idxs.size()) {
+        throw std::runtime_error("atom indices must be unique");
+    }
+    if (*std::max_element(atom_idxs.begin(), atom_idxs.end()) >= N_) {
+        throw std::runtime_error("indices values must be less than N");
+    }
+    if (*std::min_element(atom_idxs.begin(), atom_idxs.end()) < 0) {
+        throw std::runtime_error("indices values must be greater or equal to zero");
+    }
+}
+
+// Set atom idxs upon which to compute the non-bonded potential. This will trigger a neighborlist rebuild.
+template <typename RealType> void NonbondedAllPairs<RealType>::set_atom_idxs(const std::vector<int> &atom_idxs) {
+    this->verify_atom_idxs(atom_idxs);
+    const cudaStream_t stream = static_cast<cudaStream_t>(0);
+    std::vector<unsigned int> unsigned_idxs = std::vector<unsigned int>(atom_idxs.begin(), atom_idxs.end());
+    DeviceBuffer<unsigned int> atom_idxs_buffer(atom_idxs.size());
+    atom_idxs_buffer.copy_from(&unsigned_idxs[0]);
+    this->set_atom_idxs_device(atom_idxs.size(), atom_idxs_buffer.data, stream);
+    gpuErrchk(cudaStreamSynchronize(stream));
+}
+
+template <typename RealType>
+void NonbondedAllPairs<RealType>::set_atom_idxs_device(
+    const int K, const unsigned int *d_in_atom_idxs, const cudaStream_t stream) {
+    if (K < 1) {
+        throw std::runtime_error("K must be at least 1");
+    }
+    if (K > N_) {
+        throw std::runtime_error("number of idxs must be less than or equal to N");
+    }
+    gpuErrchk(
+        cudaMemcpyAsync(d_atom_idxs_, d_in_atom_idxs, K * sizeof(*d_atom_idxs_), cudaMemcpyDeviceToDevice, stream));
+    nblist_.resize_device(K, stream);
+    // Force the rebuild of the nblist
+    gpuErrchk(cudaMemsetAsync(d_rebuild_nblist_, 1, 1 * sizeof(*d_rebuild_nblist_), stream));
+    this->K_ = K;
+}
+
 template <typename RealType> void NonbondedAllPairs<RealType>::disable_hilbert_sort() { disable_hilbert_ = true; }
 
 template <typename RealType>
 void NonbondedAllPairs<RealType>::hilbert_sort(const double *d_coords, const double *d_box, cudaStream_t stream) {
 
-    const int tpb = 32;
+    const int tpb = warp_size;
     const int B = ceil_divide(K_, tpb);
 
-    if (d_atom_idxs_) {
-        k_coords_to_kv_gather<<<B, tpb, 0, stream>>>(
-            K_, d_atom_idxs_, d_coords, d_box, d_bin_to_idx_, d_sort_keys_in_, d_sort_vals_in_);
-    } else {
-        // N_ == K_
-        k_coords_to_kv<<<B, tpb, 0, stream>>>(K_, d_coords, d_box, d_bin_to_idx_, d_sort_keys_in_, d_sort_vals_in_);
```

Comment:
Can `k_coords_to_kv` be removed? (then might also rename `k_coords_to_kv{_gather,}`)

Reply:
Removed here: https://github.com/proteneer/timemachine/pull/915/commits/89fb1d8e41e5b6c9df4fb521da1199dce1d023e9

Not clear what you mean by rename. If you mean to rename `k_coords_to_kv_gather-> k_coords_to_kv`, seems like the convention of `scatter` and `gather` in the signature is nice.

Reply:
> If you mean to rename k_coords_to_kv_gather-> k_coords_to_kv

Yep, that's what I meant. Was thinking to remove the suffix since it's now the only variant. But I can see the utility of keeping `gather` in the name.